### ISSUE TITLE
About specs.

### ISSUE THREAD

**Issue by maethor** [Boavizta/Energizta#14]
I think we did not discussed enough about specs. So, after last friday's hackaton, this is what I understood and what I propose.

### Constraints

- We limit ourselves to Linux compatibility. Windows is not in scope for now.
- We only manage baremetal servers. VPS are out of scope.
- We want something than can be run easily by anyone on any server (sadly root access is needed to get RAPL and DCMI data).
    - It should be easily reviewable, because users could run it, in root, on production servers
    - It should have everything embedded in one script or one executable. Something that can be easily copy-paste or ever `curl … | bash`.
    - It should not need dependency, or dependencies must be limited to something that can be installed from Debian repository. The user should not need to pip install stuff, or to run docker, or whatever except tools that are strictly needed and can be trusted by users.

### Functions

I believe the stress test should be an option. We do not want to only get stress test data, but also regular states and power usage levels on normal server usage. The important thing is to have a collection of various states (load + power usage) on every hardware configuration imaginable. What's really running on the machine should not be important, only states (load + power usage) matter. Please tell me if I am wrong about this. 

- By default, I think the main script should enter an infinite loop and just output json lines every INTERVAL seconds on the standard output. Each json line will contain a state (load, temp, cpu, mem…) and all powers it could find (rapl, dcmi, sensors, etc.). The format is discussed here #9 
- This script could be simply used in a systemd service, and the output could be redirected to a log file for further study or monitoring purposes.

#### Stress-tests

- An option will allow run a stress-test. With this option the user specify a file that contains a list of stress-tests shell commands, one per line. Typically, a stress-test example using a series of stress-ng commands would be provided in the project repository. For each command line of the stress-test file, the script will run the stess-test while running the function that computes and display the state json every X seconds. This way, the user can easily define the stress-tests he wants to run.
- At the end of the stress-test, the script could ask the user if he/she wants to send the result to boavizta. If yes, it will add the servers specs to the json, as defined in #9 

### Ideas

- An option in seconds to let the server adapt after starting a stress-test and before starting to register states. For example the script could start a 2 minute stress-test, wait for 1 minute, and register stats only on 2nd minute.
- An option to choose between displaying states every INTERVAL seconds during the stress-tests, or tell the script to still get states every INTERVAL seconds, but only display an average at the end of each stress-test.
- An option to allow manual input of the power consumption by the user at the end of each stress-test, for testing equipment with wattmeters.
- To write in the Boavizta database, we should provide a webui and a API on Boavizta servers to allow pasting or uploading test files. We should not have complicated push stuff into the script. In the first version, we should display on the standard output and document how to redirect it to a file and how to send this file on the Boavizta database using curl.
- I the final version, the script should simplify this process, but only:
    - display a json and just ask the user to copy/paste it to a given Boavizta URL
    - register the json into a file and ask the user to upload this file to a given Boavizta URL
    - make the HTTP PUT directly, if the user is OK after seeing the json.

What do you think?
I am not sure I am able to write this in Python, but I can do it in Bash.

**Comment by da-ekchajzer** [comment]
I agree with the specs mentioned.

I would add the identification of global characteristics (which doesn't change from one state to another) : 
* Server specs (cpu, ram, gpu, disks, ...)
* Linux distribution
* Specific configuration (overclocking, ...)
* ...

I'm still a little hesitant about the data export part.

**Comment by maethor** [comment]
Of course we need global specs, but not on every "states" line. I already fear that the json output will quickly become pretty big for those who will run the script fulltime in the background (that's what I want to do).
